MobiusCore/Source/EffectHandlers/EffectHandler.swift

```diff
@@ -0,0 +1,140 @@
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+import Foundation
+
+public enum HandleEffect<Type> {
+    case handle(Type)
+    case ignore
+}
+
+/// An `EffectHandler` is a building block in Mobius loops which carry out side-effects in response to effects emitted by the `update` function.
+/// An `EffectHandler` decides which effects it can handle based on its `canAccept` function. Multiple `EffectHandler`s can be composed by using an
+/// `EffectRouterBuilder`.
+///
+/// Note: When using an `EffectRouterBuilder` each effect must be handled by exactly one `EffectHandler`.
+/// Note: The `connnect` function is invoked on an `EffectHandler` when it should start handling effects and emitting events. Only one `Connection` at
+/// a time is supported, otherwise it will crash.
+/// Note: It is possible to emit events before `connect` has been called on an `EffectHandler`, and after a `Connection` to an `EffectHandler` has
+/// been disposed. These events can not, and will not be handled by anything. This will therefore cause a crash.
+/// Note: Any resources used by an `EffectHandler` must be disposed when a connection to the `EffectHandler` is disposed. The `onDispose`
+/// parameter is used to specify which resources should be torn down when this happens.
+final public class EffectHandler<Effect, Event> {
+    private let lock = Lock()
+    private let handleEffect: (Effect) -> ((@escaping Consumer<Event>) -> Void)?
+    private let disposeFn: () -> Void
+    private var consumer: Consumer<Event>?
+
+    /// Create a handler for effects which satisfy the `canAccept` parameter function.
+    ///
+    /// - Parameter canHandle: A function which indicates whether to handle an effect. If it returns `.handle(effect)` for a given `effect`, this
+    /// effect handler will handle said effect.
+    /// - Parameter handleEffect: Handle effects which satisfy `canAccept`.
+    /// - Parameter onDispose: Tear down any resources being used by this effect handler.
+    public init<AssociatedValueType>(
+        canHandle: @escaping (Effect) -> HandleEffect<AssociatedValueType>,
+        handleEffect: @escaping (AssociatedValueType, @escaping Consumer<Event>) -> Void,
+        onDispose disposable: @escaping () -> Void
```

Comment:
On an unrelated note, we need to be very clear with documenting this `onDispose`, since it binds to the lifetime of a `connection` rather than the `connectable` (composite event handler), and may be called multiple times per event handler instance. (This distinction is why I called it `stop` instead of `dispose`)

Reply:
I think I prefer `can` over `should`, but otherwise I like the second form.

Reply:
Ok, I like it too. I'll change it :)